apps/transport/lib/transport/gbfs_metadata.ex

```diff
@@ -0,0 +1,228 @@
+defmodule Transport.GBFSMetadata do
+  @moduledoc """
+  Compute and store metadata for GBFS resources.
+  """
+  alias DB.{Dataset, Repo, Resource}
+  alias Shared.Validation.GBFSValidator.Summary, as: GBFSValidationSummary
+  alias Shared.Validation.GBFSValidator.Wrapper, as: GBFSValidator
+  import Ecto.Query
+  require Logger
+
+  @doc """
+  It is a bit of work, currently, to extract the list of `gbfs.json` endpoints,
+  for instance because `format` is not enough alone to filter them.
+
+  See https://github.com/etalab/transport-site/issues/1891#issuecomment-958888421 for some background.
+  """
+  def gbfs_feeds_query do
+    Resource
+    |> join(:inner, [r], d in Dataset, on: r.dataset_id == d.id)
+    |> where([_r, d], d.type == "bike-scooter-sharing" and d.is_active)
+    |> where([r, _d], like(r.url, "%gbfs.json") or r.format == "gbfs")
+    |> where([r, _d], not fragment("? ~ ?", r.url, "station|free_bike"))
+  end
+
+  def set_gbfs_feeds_metadata do
+    resources = gbfs_feeds_query() |> Repo.all()
+
+    Logger.info("Fetching details about #{Enum.count(resources)} GBFS feeds")
+
+    resources
+    |> Stream.map(fn resource ->
+      Logger.info("Fetching GBFS metadata for #{resource.url} (##{resource.id})")
+      changeset = Resource.changeset(resource, %{format: "gbfs", metadata: compute_feed_metadata(resource)})
+      Repo.update!(changeset)
+    end)
+    |> Stream.run()
+  end
+
+  @doc """
+  This function does 2 HTTP calls on a given resource url, and returns a report
+  with metadata and also validation status (using a third-party HTTP validator).
+  """
+  @spec compute_feed_metadata(Resource.t()) :: map()
+  def compute_feed_metadata(resource) do
+    with {:ok, %{status_code: 200, body: body} = response} <-
+           http_client().get(resource.url, [{"origin", website_url()}]),
+         {:ok, json} <- Jason.decode(body) do
+      %{
+        validation: validation(resource),
+        has_cors: has_cors?(response),
+        is_cors_allowed: cors_headers_allows_self?(response),
+        feeds: feeds(json),
+        versions: versions(json),
+        languages: languages(json),
+        system_details: system_details(json),
+        types: types(json),
+        ttl: ttl(json)
+      }
+    else
+      e ->
+        Logger.error(inspect(e))
+        %{}
+    end
+  end
+
+  @spec validation(Resource.t()) :: GBFSValidationSummary.t() | nil
+  defp validation(resource) do
+    case GBFSValidator.validate(resource.url) do
+      {:ok, %GBFSValidationSummary{} = summary} -> summary
+      {:error, _} -> nil
+    end
+  end
+
+  defp types(%{"data" => _data} = payload) do
```

Comment:
Dans un style un peu plus fonctionnel, je suggère :
```suggestion
  defp types(%{"data" => _data} = payload) do
    result =
      []
      |> maybe_add_type("free_floating", has_feed?(payload, "free_bike_status"))
      |> maybe_add_type("stations", has_feed?(payload, "station_information"))

    if Enum.empty?(result) do
      Logger.error("Cannot detect GBFS types for feed #{inspect(payload)}")
      nil
    else
      result
    end
  end

  def maybe_add_type(result, tag, condition) do
    if condition, do: result ++ [tag], else: result
  end
```

Reply:
(à vérifier!)